### PR TITLE
add error log for python api and c++ 

### PR DESCRIPTION
**Pull request by ceci3**
add conv,dropout,batch_norm, cross_entropy c++ shape log

### **1、 conv ：**
  **1.1** 
  `x=fluid.layers.data(name='x', append_batch_size = False, shape=[2, 5], dtype='float64')
  res = fluid.layers.conv2d(x, 3, 3)`
 修改前报错：
 ` Conv intput should be 4-D or 5-D tensor, get 2`
 修改后报错：
  `ShapeError: Conv intput should be 4-D or 5-D tensor, But received: 2-D Tensor.`
  **1.2**
  `x=fluid.layers.data(name='x', append_batch_size = False, shape=[2, 5, 5, 6], dtype='float64')
  res = fluid.layers.conv2d(x, 3, 1, groups=2)`
  修改前报错：
  `ValueError: num_channels must be divisible by groups.`
  修改后报错：
  `ValueError:  the channel of input must be divisible by groups， received: the channel of input is 5, the shape of input is [2, 5, 5, 6], the groups is 2`

### **2、batch_norm** 
`x=fluid.layers.data(name='x', append_batch_size = False, shape=[2, 5, 5, 6, 7, 7], dtype='float64')
res = fluid.layers.batch_norm(x)`
修改前报错：
`Input X must have 2 to 5 dimensions.`
修改后报错：
`ShapeError: Input X must have 2 to 5 dimensions. But received:the dimensions of Input X is [6]`

### **3、cross_entropy**
3.1 
`x=fluid.layers.data(name='x', append_batch_size = False, shape=[2, 5], dtype='float64')
lab=fluid.layers.data(name='lab', append_batch_size = False, shape=[5, 3], dtype='float64')
res = fluid.layers.cross_entropy(x, lab)`
修改前报错：
`Input(X) and Input(Label) shall have the same shape except the last dimension.`
修改后报错：
`ShapeError: Input(X) and Input(Label) shall have the same shape except the last dimension. But received: the shape of Input(X) is [2, 5],the shape of Input(Label) is [5, 3].`
3.2 
`x=fluid.layers.data(name='x', append_batch_size = False, shape=[2, 5], dtype='float64')
lab=fluid.layers.data(name='lab', append_batch_size = False, shape=[2, 3], dtype='float64')
res = fluid.layers.cross_entropy(x, lab)`
修改前报错：
`the last dimension of Input(Label) should be 1.`
修改后报错：
`ShapeError: the last dimension of Input(Label) should be 1.But received: the last dimension of Input(Label) is [3], the last dimension is [1].`
3.3
`x=fluid.layers.data(name='x', append_batch_size = False, shape=[2, 5], dtype='float64')
lab=fluid.layers.data(name='lab', append_batch_size = False, shape=[2, 3], dtype='float64')
res = fluid.layers.cross_entropy(x, lab, soft_label=True)`
修改前报错：
`If Attr(soft_label) == true, the last dimension of Input(X) and Input(Label) should be equal.`
修改后报错：
`ShapeError: If Attr(soft_label) == true, the last dimension of Input(X) and Input(Label) should be equal. But received: the last dimension of Input(X) is [5], the shape of Input(X) is [2,5], the last dimension of Input(Label) is [3], the shape of Input(Label) is [2,3], the last dimension is [1].`

### CODE REVIEW

File: paddle/fluid/operators/batch_norm_op.cc [hunk 1]
```diff
@@ -59,7 +59,8 @@ void BatchNormOp::InferShape(framework::InferShapeContext *ctx) const {
       ctx->Attrs().Get<std::string>("data_layout"));
 
   PADDLE_ENFORCE(x_dims.size() >= 2 && x_dims.size() <= 5,
-                 "Input X must have 2 to 5 dimensions.");
+                 "ShapeError: Input X must have 2 to 5 dimensions. But received:"
+                 "the dimensions of Input X is [%d]", x_dims.size());
```
Comment (luotao1) [inline paddle/fluid/operators/batch_norm_op.cc]: 1. 请看示例，需要打出x_dims，不只是x_dims.size()。下同，所有的dims都需要打出来。
2. 请将PADDLE_ENFORCE条件语句拆成2个，不要有&&。

File: paddle/fluid/operators/batch_norm_op.cc [hunk 2]
```diff
@@ -78,8 +81,10 @@ void BatchNormOp::InferShape(framework::InferShapeContext *ctx) const {
   }
 
   if (check) {
-    PADDLE_ENFORCE_EQ(scale_dim[0], C);
-    PADDLE_ENFORCE_EQ(scale_dim[0], C);
+    PADDLE_ENFORCE_EQ(scale_dim[0], C, "ShapeError: the shape of scale must equal to [%d]"
```
Comment (luotao1) [inline paddle/fluid/operators/batch_norm_op.cc]: shape of scale 这是scale_dim[0]的意思么？bias_dim[0]同

File: paddle/fluid/operators/conv_op.cc [hunk 3]
```diff
@@ -46,26 +46,35 @@ void ConvOp::InferShape(framework::InferShapeContext* ctx) const {
   int groups = ctx->Attrs().Get<int>("groups");
   std::vector<int> dilations = ctx->Attrs().Get<std::vector<int>>("dilations");
 
-  PADDLE_ENFORCE(in_dims.size() == 4 || in_dims.size() == 5,
-                 "Conv intput should be 4-D or 5-D tensor, get %u",
+  PADDLE_ENFORCE_EQ(in_dims.size() == 4 || in_dims.size() == 5, true,
+                 "ShapeError: Conv intput should be 4-D or 5-D tensor, But received: %u-D Tensor",
```
Comment (luotao1) [inline paddle/fluid/operators/conv_op.cc]: 逗号后，But要小写。

File: paddle/fluid/operators/conv_op.cc [hunk 4]
```diff
@@ -46,26 +46,35 @@ void ConvOp::InferShape(framework::InferShapeContext* ctx) const {
...
-      "Conv input dimension and filter dimension should be the same.");
-  PADDLE_ENFORCE(
-      in_dims.size() - strides.size() == 2U,
-      "Conv input dimension and strides dimension should be consistent.");
+      "ShapeError: Conv input dimension and filter dimension should be the same."
+      "But received: input dimension of Conv is [%d], the filter dimension" 
+      "of Conv is [%d]", in_dims.size(), filter_dims.size());
+  PADDLE_ENFORCE_EQ(
+      in_dims.size() - strides.size() == 2U, true,
+      "ShapeError: Conv input dimension and strides dimension should be consistent."
```
Comment (luotao1) [inline paddle/fluid/operators/conv_op.cc]: Conv input dimension and strides dimension should be consistent。这句和in_dims.size() - strides.size() == 2U不太符合吧

File: paddle/fluid/operators/batch_norm_op.cc (hunk 1 above)
Comment (ceci3) [reply paddle/fluid/operators/batch_norm_op.cc]: Done， thanks.

File: paddle/fluid/operators/conv_op.cc (hunk 3 above)
Comment (ceci3) [reply paddle/fluid/operators/conv_op.cc]: Done, thanks

File: paddle/fluid/operators/conv_op.cc (hunk 4 above)
Comment (ceci3) [reply paddle/fluid/operators/conv_op.cc]: Done, thanks

File: paddle/fluid/operators/batch_norm_op.cc (hunk 2 above)
Comment (ceci3) [reply paddle/fluid/operators/batch_norm_op.cc]: 是的，scale和bias都是一维的，所以直接判断scale_dim[0]和bias_dim[0] 是否和channel相等就行。

Comment (luotao1) [reply paddle/fluid/operators/batch_norm_op.cc]: 这里是说scale_dim和bias_dim都只有1维么？

File: paddle/fluid/operators/conv_op.cc [hunk 5]
```diff
@@ -46,26 +46,39 @@ void ConvOp::InferShape(framework::InferShapeContext* ctx) const {
...
+      in_dims, in_dims.size(), filter_dims, filter_dims.size());
+
+  int in_sub_stride_size = in_dims.size() - strides.size();
+  PADDLE_ENFORCE_EQ(
+      in_dims.size() - strides.size() == 2U, true,
+      "ShapeError: the dimension of input minus the dimension of stride must euqal to 2."
+      "But received: the dimension of input minus the dimension of stride is [%d], the"
+      "input dimension of Conv is [%d], the shape of Conv input is [%s], the stride"
+      "dimension of Conv is [%d]",
+      in_sub_stride_size, in_dims.size(), in_dims, strides.size());
```
Comment (luotao1) [inline paddle/fluid/operators/conv_op.cc]: strides和paddings的维度信息需要打出来。72行同。

File: paddle/fluid/operators/conv_op.cc [hunk 6]
```diff
@@ -46,26 +46,39 @@ void ConvOp::InferShape(framework::InferShapeContext* ctx) const {
...
+      "ShapeError: Conv paddings dimension and Conv strides dimension should be the same."
+      "But received: The paddings dimension of Conv is [%d]"
+      "the stride dimension of Conv is [%d]", paddings.size(), strides.size());
 
   PADDLE_ENFORCE_EQ(in_dims[1], filter_dims[1] * groups,
-                    "The number of input channels should be equal to filter "
-                    "channels * groups.");
+                    "ShapeError: The number of input channels should be equal to filter "
+                    "channels * groups. But received: the input channels is [%d], the filter channel is [%d],"
+                    "the groups is [%d]", in_dims[1], filter_dims[1], groups);
```
Comment (luotao1) [inline paddle/fluid/operators/conv_op.cc]: in_dims和filter_dims的具体维度信息要打出来。80行同。

File: paddle/fluid/operators/cross_entropy_op.cc [hunk 7]
```diff
@@ -43,28 +43,38 @@ class CrossEntropyOpBase : public framework::OperatorWithKernel {
...
+          "the shape of Input(X) is [%s], the dimensions of Input(Label) is [%d], the shape of"
+          "Input(Label) is [%s]", rank, x_dims, label_dims.size(), label_dims);
       if (check) {
         PADDLE_ENFORCE_EQ(x_dims[rank - 1], label_dims[rank - 1],
-                          "If Attr(soft_label) == true, the last dimension of "
-                          "Input(X) and Input(Label) should be equal.");
+                          "ShapeError: If Attr(soft_label) == true, the last dimension of "
+                          "Input(X) and Input(Label) should be equal. But received: the" 
+                          "last dimension of Input(X) is [%d], the last dimension of"
+                          "Input(Label) is [%d].", x_dims[rank - 1], label_dims[rank - 1]);
```
Comment (luotao1) [inline paddle/fluid/operators/cross_entropy_op.cc]: 需要打出x_dims和label_dims的维度信息，以及说明rank-1是第几维。69行同。

File: paddle/fluid/operators/batch_norm_op.cc [hunk 8]
```diff
@@ -68,8 +74,12 @@ void BatchNormOp::InferShape(framework::InferShapeContext *ctx) const {
   auto scale_dim = ctx->GetInputDim("Scale");
   auto bias_dim = ctx->GetInputDim("Bias");
 
-  PADDLE_ENFORCE_EQ(scale_dim.size(), 1UL);
-  PADDLE_ENFORCE_EQ(scale_dim.size(), 1UL);
+  PADDLE_ENFORCE_EQ(scale_dim.size(), 1UL, "ShapeError: the dimension of scale must equal to 1."
+                 "But received: the shape of scale is [%s], the dimension of scale is [%d]",
+                 scale_dim, scale_dim.size());
+  PADDLE_ENFORCE_EQ(bias_dim.size(), 1UL, "ShapeError: the dimension of bias must equal to 1."
+                 "But received: the shape of scale is [%s],the dimension of scale is [%d]",
```
Comment (chenwhql) [inline paddle/fluid/operators/batch_norm_op.cc]: `scale` -> `bias`

File: paddle/fluid/operators/batch_norm_op.cc [hunk 9]
```diff
@@ -78,8 +88,10 @@ void BatchNormOp::InferShape(framework::InferShapeContext *ctx) const {
   }
 
   if (check) {
-    PADDLE_ENFORCE_EQ(scale_dim[0], C);
-    PADDLE_ENFORCE_EQ(scale_dim[0], C);
+    PADDLE_ENFORCE_EQ(scale_dim[0], C, "ShapeError: the shape of scale must equal to [%d]"
```
Comment (chenwhql) [inline paddle/fluid/operators/batch_norm_op.cc]: `shape`是一个整体概念，`C`是一个int值，这里比较是不是不在一个层面？

File: paddle/fluid/operators/conv_op.cc [hunk 10]
```diff
@@ -46,26 +46,39 @@ void ConvOp::InferShape(framework::InferShapeContext* ctx) const {
...
+                 "ShapeError: Conv intput should be 4-D or 5-D tensor. But received: %u-D Tensor,"
+                 "the shape of Conv input is [%s]", in_dims.size(), in_dims);
 
   PADDLE_ENFORCE_EQ(
       in_dims.size(), filter_dims.size(),
-      "Conv input dimension and filter dimension should be the same.");
-  PADDLE_ENFORCE(
-      in_dims.size() - strides.size() == 2U,
-      "Conv input dimension and strides dimension should be consistent.");
+      "ShapeError: Conv input dimension and filter dimension should be the same."
```
Comment (chenwhql) [inline paddle/fluid/operators/conv_op.cc]: `should be the same`有点别扭，`should be equal`?

File: paddle/fluid/operators/conv_op.cc [hunk 11]
```diff
@@ -46,26 +46,39 @@ void ConvOp::InferShape(framework::InferShapeContext* ctx) const {
...
+      "But received: the shape of Conv input is [%s], input dimension of Conv input is [%d],"
+      "the shape of filter is [%s],  the filter dimension of Conv is [%d]",
+      in_dims, in_dims.size(), filter_dims, filter_dims.size());
+
+  int in_sub_stride_size = in_dims.size() - strides.size();
+  PADDLE_ENFORCE_EQ(
+      in_dims.size() - strides.size() == 2U, true,
+      "ShapeError: the dimension of input minus the dimension of stride must euqal to 2."
```
Comment (chenwhql) [inline paddle/fluid/operators/conv_op.cc]: `must be equal`?

File: paddle/fluid/operators/conv_op.cc [hunk 12]
```diff
@@ -46,26 +46,39 @@ void ConvOp::InferShape(framework::InferShapeContext* ctx) const {
...
+      "But received: the dimension of input minus the dimension of stride is [%d], the"
+      "input dimension of Conv is [%d], the shape of Conv input is [%s], the stride"
+      "dimension of Conv is [%d]",
+      in_sub_stride_size, in_dims.size(), in_dims, strides.size());
   PADDLE_ENFORCE_EQ(
       paddings.size(), strides.size(),
-      "Conv paddings dimension and Conv strides dimension should be the same.");
+      "ShapeError: Conv paddings dimension and Conv strides dimension should be the same."
```
Comment (chenwhql) [inline paddle/fluid/operators/conv_op.cc]: 同上，`should be the same`改一下？

**Review by chenwhql — COMMENTED** [review]
可以及时更新下PR的comment描述

File: paddle/fluid/operators/batch_norm_op.cc (hunk 2 above)
Comment (ceci3) [reply paddle/fluid/operators/batch_norm_op.cc]: 是的，scale_dim和bias_dim都是一维，scale和bias大小都应该是输入的通道[C]

File: paddle/fluid/operators/conv_op.cc (hunk 6 above)
Comment (ceci3) [reply paddle/fluid/operators/conv_op.cc]: 这里对比的是第一维度的channel，打印出全部的维度信息会不会有点冗余。

File: paddle/fluid/operators/cross_entropy_op.cc (hunk 7 above)
Comment (ceci3) [reply paddle/fluid/operators/cross_entropy_op.cc]: op文档里说明了最后一个维度要保持相等或者为1，错误信息里说明了对比的是是最后一个维度的值，rank-1在打出来一个数字感觉会造成疑惑。 69行是一样的情况。

File: paddle/fluid/operators/conv_op.cc (hunk 5 above)
Comment (ceci3) [reply paddle/fluid/operators/conv_op.cc]: 这两个是vector的属性，所以就没打维度信息

File: paddle/fluid/operators/batch_norm_op.cc (hunk 9 above)
Comment (ceci3) [reply paddle/fluid/operators/batch_norm_op.cc]: 但是scale其实是一个1维的，所以scale里也只有一个int类型的值，我在打印的C的前后添加了[]，打出是不是可以看做是一个shape？

File: paddle/fluid/operators/conv_op.cc (hunk 12 above)
Comment (ceci3) [reply paddle/fluid/operators/conv_op.cc]: done，thanks

File: paddle/fluid/operators/conv_op.cc (hunk 10 above)
Comment (ceci3) [reply paddle/fluid/operators/conv_op.cc]: done，thanks

File: paddle/fluid/operators/conv_op.cc (hunk 11 above)
Comment (ceci3) [reply paddle/fluid/operators/conv_op.cc]: done，thanks

File: paddle/fluid/operators/batch_norm_op.cc (hunk 8 above)
Comment (ceci3) [reply paddle/fluid/operators/batch_norm_op.cc]: done，thanks

File: paddle/fluid/operators/conv_op.cc (hunk 6 above)
Comment (luotao1) [reply paddle/fluid/operators/conv_op.cc]: 并不会冗余。这样方便用户知道全局的维度信息。

File: paddle/fluid/operators/cross_entropy_op.cc (hunk 7 above)
Comment (luotao1) [reply paddle/fluid/operators/cross_entropy_op.cc]: 是要说明最后一个维度，是第几维。

File: paddle/fluid/operators/batch_norm_op.cc (hunk 9 above)
Comment (luotao1) [reply paddle/fluid/operators/batch_norm_op.cc]: 如果scale是一维的话，可以“我在打印的C的前后添加了[]，打出是不是可以看做是一个shape”

File: paddle/fluid/operators/cross_entropy_op.cc (hunk 7 above)
Comment (ceci3) [reply paddle/fluid/operators/cross_entropy_op.cc]: done, thanks

File: paddle/fluid/operators/conv_op.cc (hunk 6 above)
Comment (ceci3) [reply paddle/fluid/operators/conv_op.cc]: done, thanks

**Review by chenwhql — DISMISSED** [review]
LGTM

**Review by chenwhql — DISMISSED** [review]

**Review by luotao1 — APPROVED** [review]

**Pull request merged**